pam/integration-tests/integration_test.go

```diff
@@ -0,0 +1,181 @@
+	addEnv := []string{
+		"TESTS_GPASSWD_ARGS=" + strings.Join(gpasswdArgs, "-sep-"),
+		"TESTS_GPASSWD_GRP_FILE_PATH=" + gpasswdFile,
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	_, stopped := testutils.RunDaemon(ctx, t, daemonPath, testutils.WithSocketPath("/tmp/pam-cli-tests.sock"), testutils.WithEnvironment(addEnv...))
+	t.Cleanup(func() {
+		cancel()
+		<-stopped
+	})
+
+	// If vhs is installed with "go install", we need to add GOPATH to PATH.
+	pathEnv := appendGoBinToPath(t)
+
+	currentDir, err := os.Getwd()
+	require.NoError(t, err, "Setup: Could not get current directory for the tests")
+
+	tests := map[string]struct {
+		tape string
+	}{
+		"Authenticate user successfully":               {tape: "simple_auth"},
+		"Authenticate user with mfa":                   {tape: "mfa_auth"},
+		"Authenticate user with form mode with button": {tape: "form_with_button"},
+		"Authenticate user with qr code":               {tape: "qr_code"},
+		"Authenticate user and reset password":         {tape: "mandatory_password_reset"},
+		"Authenticate user and offer password reset":   {tape: "optional_password_reset"},
+		"Authenticate user switching auth mode":        {tape: "switch_auth_mode"},
+		"Authenticate user switching username":         {tape: "switch_username"},
+		"Authenticate user switching broker":           {tape: "switch_broker"},
+		"Authenticate user and add it to local group":  {tape: "local_group"},
+
+		"Remember last successful broker and mode": {tape: "remember_broker_and_mode"},
+
+		"Deny authentication if max attempts reached": {tape: "max_attempts"},
+		"Deny authentication if user does not exist":  {tape: "unexistent_user"},
+
+		"Exit authd if local broker is selected": {tape: "local_broker"},
+		"Exit authd if user sigints":             {tape: "sigint"},
+	}
+	for name, tc := range tests {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			// #nosec:G204 - we control the command arguments in tests
+			cmd := exec.Command("vhs", filepath.Join(currentDir, "testdata", "tapes", tc.tape+".tape"))
+			cmd.Env = testutils.AppendCovEnv(cmd.Env)
+			cmd.Env = append(cmd.Env, pathEnv)
+			cmd.Dir = outDir
+
+			out, err := cmd.CombinedOutput()
+			require.NoError(t, err, "Failed to run tape %q: %v: %s", tc.tape, err, out)
+
+			tmp, err := os.ReadFile(filepath.Join(outDir, tc.tape+".txt"))
+			require.NoError(t, err, "Could not read output file of tape %q", tc.tape)
+
+			// We need to format the output a little bit, since the txt file can have some noise at the beginning.
+			var got string
+			splitTmp := strings.Split(string(tmp), "\n")
+			for i, str := range splitTmp {
+				if strings.HasPrefix(str, "> ./pam_authd socket=/tmp/pam-cli-tests.sock") {
+					got = strings.Join(splitTmp[i:], "\n")
+					break
+				}
+			}
+			want := testutils.LoadWithUpdateFromGolden(t, got)
+			require.Equal(t, want, got, "Output of tape %q does not match golden file", tc.tape)
+
+			if tc.tape == "local_group" {
```

Comment:
I'm not in love with having test-specific behaviors inside the test itself, wouldn't be better to be this defined inside a function that is defined together with the table of tests?

Reply:
That would help to clear the test function itself, indeed. But it would affect the test case readability. 

Having clarity in the test case definition is way better than having clarity in the test function, since we look and alter them way more consistently than the test function itself.